var/spack/repos/builtin/packages/gasnet/package.py

```diff
@@ -6,98 +6,133 @@
-    version('1.28.2', sha256='7903fd8ebdd03bcda20a66e3fcedef2f8b384324591aa91b8370f3360f6384eb')
-    version('1.28.0', sha256='a7999fbaa1f220c2eb9657279c7e7cccd1b21865d5383c9a5685cfe05a0702bc')
-    version('1.24.0', sha256='76b4d897d5e2261ef83d0885c192e8ac039e32cb2464f11eb64eb3f9f2df38c0')
-
-    variant('mpi', default=True, description="Support MPI")
-    variant('ibv', default=False, description="Support InfiniBand")
-    variant('udp', default=False, description="Support UDP")
-    variant('aligned-segments', default=False,
-            description="Requirement to achieve aligned VM segments")
-    variant('pshm', default=True,
-            description="Support inter-process shared memory support")
-    variant('segment-mmap-max', default='16GB',
-            description="Upper bound for mmap-based GASNet segments")
-
-    conflicts('+aligned-segments', when='+pshm')
+    version('develop', branch='develop')
+    version('main',    branch='stable')
+    version('master',  branch='master')
 
-    depends_on('mpi', when='+mpi')
+    version('2021.3.0',  sha256='8a40fb3fa8bacc3922cd4d45217816fcb60100357ab97fb622a245567ea31747')
+    version('2020.10.0', sha256='ed17baf7fce90499b539857ee37b3eea961aa475cffbde77e4c607a34ece06a0')
+    version('2020.3.0',  sha256='019eb2d2284856e6fabe6c8c0061c874f10e95fa0265245f227fd3497f1bb274')
+    version('2019.9.0',  sha256='117f5fdb16e53d0fa8a47a1e28cccab1d8020ed4f6e50163d985dc90226aaa2c')
+    # Do NOT add older versions here.
+    # GASNet-EX releases over 2 years old are not supported.
 
-    def url_for_version(self, version):
-        url = "https://gasnet.lbl.gov/"
-        if version >= Version('2019'):
-            url += "EX/GASNet-{0}.tar.gz".format(version)
-        else:
-            url += "download/GASNet-{0}.tar.gz".format(version)
-
-        return url
-
-    def configure_args(self):
-        args = [
-            # TODO: factor IB suport out into architecture description.
-            '--enable-par',
-            '--enable-mpi-compat',
-            '--enable-segment-fast',
-            '--disable-parsync',
-            '--with-segment-mmap-max=%s '
-            % (self.spec.variants['segment-mmap-max'].value),
-            # for consumers with shared libs
-            "CC=%s %s" % (spack_cc, self.compiler.cc_pic_flag),
-            "CXX=%s %s" % (spack_cxx, self.compiler.cxx_pic_flag),
-        ]
-
-        if '+aligned-segments' in self.spec:
-            args.append('--enable-aligned-segments')
-        else:
-            args.append('--disable-aligned-segments')
+    # The optional network backends:
+    variant('smp', default=True, description="SMP conduit for single-node operation")
+    variant('mpi', default=False, description="Low-performance/portable MPI conduit")
+    variant('ibv', default=False, description="Native InfiniBand verbs conduit")
+    variant('udp', default=False, description="Portable UDP conduit, for Ethernet networks")
```

Comment:
This should be a single variant
```
variant('conduits', values=any_combination_of('smp', 'mpi', 'ibv', 'udp').with_default('smp'), 
             description=...)
```

Reply:
This would be a major departure from the variants in the package this is replacing, but I'm fine implementing that if this is the consensus view?

Reply:
IIUC One significant downside of this suggestion is it would narrow/limit the amount of user documentation the package can conveniently provide for each value in this most critical setting.

Current `spack info gasnet`:
```
Variants:
    Name [Default]    Allowed values    Description
    ==============    ==============    ===========================================

    debug [off]       on, off           Enable library debugging mode
    ibv [off]         on, off           Native InfiniBand verbs conduit
    mpi [off]         on, off           Low-performance/portable MPI conduit
    smp [on]          on, off           SMP conduit for single-node operation
    udp [off]         on, off           Portable UDP conduit, for Ethernet networks
```


Proposed `spack info gasnet`:

```
Variants:
    Name [Default]    Allowed values          Description
    ==============    ====================    ===========================================================================

    conduits [smp]    none, udp, smp, mpi,    The hardware-dependent network backends to enable. For detailed
                      ibv                     recommendations, consult https://gasnet.lbl.gov
    debug [off]       on, off                 Enable library debugging mode
```

A novice user has some chance of recognizing the correct/relevant network hardware backend from the first info output, but the second version sacrifices that information and might lead to a user making a less appropriate choice (that could either fail to build, fail at runtime, or fail to best leverage the available network hardware). 

Is there any way to provide better/formatted info documentation for each value in a `values=any_combination_of`?

Reply:
As @bonachea's co-developer on GASNet, I too would prefer the richer `info` output afforded by the current variants over the alternative.  When the end user who has InfiniBand hardware gets it wrong, they could easily get correct behavior from udp or mpi but _horrible_ performance.   The current descriptions go a long way toward addressing that sort of thing.

Reply:
There is no limit to the length of the docstring you give for a variant. You can add  information on every conduit option to the variant description if that will be useful for users.

Reply:
becker33 said:
> There is no limit to the length of the docstring you give for a variant

Unfortunately `spack info` appears to squash/strip all formatting whitespace, so there's no clean way to present a bulleted list of value/description pairs.

I've gone ahead and made the requested change in cae743d, although I think the resulting output is less readable:
```
Variants:
    Name [Default]    Allowed values          Description
    ==============    ====================    ==================================

    conduits [smp]    none, mpi, ibv, smp,    The hardware-dependent network
                      udp                     backends to enable. (smp) = SMP
                                              conduit for single-node operation
                                              ; (ibv) = Native InfiniBand verbs
                                              conduit ; (udp) = Portable UDP
                                              conduit, for Ethernet networks ;
                                              (mpi) = Low-performance/portable
                                              MPI conduit ; For detailed
                                              recommendations, consult
                                              https://gasnet.lbl.gov
    debug [off]       on, off                 Enable library debugging mode
```